### PR TITLE
Integrate domains via integration events

### PR DESCRIPTION
**Pull request by sven1103**
Sets up an embedded ActiveMQ Artemis server as development middle-ware for cross-domain event integration. JMS is used to abstract the underlying middle-ware implementation.

image

### CODE REVIEW

File: project-management/src/main/java/life/qbic/projectmanagement/application/policy/integration/UserRegistered.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  @Override
+  public void onReceive(IntegrationEvent event) {
+    if (!event.type().equals(TOPIC)) {
+      throw new WrongTypeException("Unknown type " + event.type());
+    }
+    var userId = Optional.ofNullable(
+        event.content().getOrDefault("userId", null));
+    System.out.println("Call JobRunner");
```
Comment (wow-such-code) [inline project-management/src/main/java/life/qbic/projectmanagement/application/policy/integration/UserRegistered.java:42]:
```suggestion
```

File: identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/broadcasting/MessageDispatcher.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * <p>
+ * Implementation of the {@link EventHub} interface, broadcasting events to the messaging
+ * middleware.
+ *
+ * @since 1.0.0
+ */
+@Component
+public class MessageDispatcher implements EventHub {
+
+  private static final String QUEUE = "User";
```
Comment (wow-such-code) [inline identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/broadcasting/MessageDispatcher.java:23]: What is the meaning/use of this String?

File: identity/src/main/java/life/qbic/identity/application/communication/broadcasting/Subscriber.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+   * @param event the integration event to receive
+   * @throws WrongTypeException if the events type returned by {@link IntegrationEvent#type()} does
+   *                            not match the subscriber's type indicated with
+   *                            {@link Subscriber#type()}
+   * @since 1.0.0
+   */
+  void onReceive(IntegrationEvent event) throws WrongTypeException;
+
+  /**
+   * Exception that indicates, that the subscriber receveived an integration event of the wrong
```
Comment (wow-such-code) [inline identity/src/main/java/life/qbic/identity/application/communication/broadcasting/Subscriber.java:42]:
```suggestion
   * Exception that indicates, that the subscriber received an integration event of the wrong
```

File: project-management/src/main/java/life/qbic/projectmanagement/application/communication/broadcasting/Subscriber.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+   * @param event the integration event to receive
+   * @throws WrongTypeException if the events type returned by {@link IntegrationEvent#type()} does
+   *                            not match the subscriber's type indicated with
+   *                            {@link Subscriber#type()}
+   * @since 1.0.0
+   */
+  void onReceive(IntegrationEvent event) throws WrongTypeException;
+
+  /**
+   * Exception that indicates, that the subscriber receveived an integration event of the wrong
```
Comment (wow-such-code) [inline project-management/src/main/java/life/qbic/projectmanagement/application/communication/broadcasting/Subscriber.java:42]:
```suggestion
   * Exception that indicates, that the subscriber received an integration event of the wrong
```

File: identity/src/main/java/life/qbic/identity/application/communication/broadcasting/IntegrationEvent.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package life.qbic.identity.application.communication.broadcasting;
+
+import java.util.Map;
+
+/**
+ * <b>Integration Event</b>
+ * <p>
+ * Integration events can be used to broadcast events from on domain to multiple other domains.
```
Comment (wow-such-code) [inline identity/src/main/java/life/qbic/identity/application/communication/broadcasting/IntegrationEvent.java:8]:
```suggestion
 * Integration events can be used to broadcast events from one domain to multiple other domains.
```

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageConsumer.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.jms.annotation.JmsListener;
+import org.springframework.stereotype.Component;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (wow-such-code) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageConsumer.java:20]: Please add a description

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageConsumer.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  private final MessageRouter messageRouter;
+
+  @Autowired
+  public MessageConsumer(MessageRouter messageRouter) {
+    this.messageRouter = messageRouter;
+    log.info("Build Message Consumer");
+  }
+
+  @JmsListener(destination = "User")
```
Comment (wow-such-code) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageConsumer.java:34]: Is this
`private static final String QUEUE = "User";`
from MessageDispatcher? Any way we can define this in one place only?

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageRouterImpl.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import life.qbic.projectmanagement.application.communication.broadcasting.Subscriber;
+import org.springframework.stereotype.Component;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (wow-such-code) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageRouterImpl.java:16]: Please add a description

File: project-management/src/main/java/life/qbic/projectmanagement/application/communication/broadcasting/MessageRouter.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
+package life.qbic.projectmanagement.application.communication.broadcasting;
+
+/**
+ * <b><interface short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (wow-such-code) [inline project-management/src/main/java/life/qbic/projectmanagement/application/communication/broadcasting/MessageRouter.java:9]:
```suggestion
```

File: user-interface/src/main/java/life/qbic/datamanager/Application.java [hunk 10]
```diff
@@ -37,6 +40,7 @@
 import org.springframework.context.ConfigurableApplicationContext;
 import org.springframework.context.annotation.ComponentScan;
 import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
+import org.springframework.jms.annotation.JmsListener;
 
 /**
  * The entry point of the Spring Boot application.
```
Comment (wow-such-code) [inline user-interface/src/main/java/life/qbic/datamanager/Application.java:46]: NIT: A lot of imports in the Application class are now unnecessary after your cleanup.

File: project-management/src/main/java/life/qbic/projectmanagement/application/policy/integration/UserRegistered.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import org.jobrunr.jobs.annotations.Job;
+import org.jobrunr.scheduling.JobScheduler;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (wow-such-code) [inline project-management/src/main/java/life/qbic/projectmanagement/application/policy/integration/UserRegistered.java:16]: Missing description

File: project-management/src/main/java/life/qbic/projectmanagement/application/communication/broadcasting/IntegrationEvent.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
+package life.qbic.projectmanagement.application.communication.broadcasting;
+
+import java.util.Map;
+
+/**
+ * <b>Integration Event</b>
+ * <p>
+ * Integration events can be used to broadcast events from on domain to multiple other domains.
```
Comment (wow-such-code) [inline project-management/src/main/java/life/qbic/projectmanagement/application/communication/broadcasting/IntegrationEvent.java:8]:
```suggestion
 * Integration events can be used to broadcast events from one domain to multiple other domains.
```

**Review by wow-such-code — CHANGES_REQUESTED** [review]
Looks very good, just a few questions/suggestions. I will also test it tomorrow morning.

File: identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/broadcasting/MessageDispatcher.java (hunk 2 above)
Comment (sven1103) [reply identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/broadcasting/MessageDispatcher.java:23]: Messages will be send to the queue/topic "User"

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageConsumer.java (hunk 7 above)
Comment (sven1103) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/communication/MessageConsumer.java:34]: Well spotted! I think it is possible to ingest this dynamically and configure it in the ``application.properties``

File: user-interface/src/main/java/life/qbic/datamanager/Application.java (hunk 10 above)
Comment (sven1103) [reply user-interface/src/main/java/life/qbic/datamanager/Application.java:46]: will clean up!

**Review by wow-such-code — APPROVED** [review]
image
Everything seems to work now!

**Pull request merged**